### PR TITLE
Refactor demand mechanism

### PR DESCRIPTION
**Pull request by FelonEkonom**
Closes https://github.com/membraneframework/membrane_core/issues/706

### CODE REVIEW

File: lib/membrane/core/element/demand_controller.ex [hunk 1]
```diff
@@ -25,8 +29,9 @@ defmodule Membrane.Core.Element.DemandController do
   def snapshot_atomic_demand(pad_ref, state) do
     with {:ok, pad_data} when not pad_data.end_of_stream? <- PadModel.get_data(state, pad_ref),
          %State{playback: :playing} <- state do
-      if pad_data.direction == :input,
-        do: raise("cannot snapshot atomic counter in input pad")
+      if pad_data.direction == :input do
+        raise("cannot snapshot atomic counter in input pad")
```
Comment (mat-hek) [inline lib/membrane/core/element/demand_controller.ex:33]: add pad ref

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Comment by mat-hek** [comment]
@FelonEkonom BTW I recall you wanted to refactor `InputQueue` as well

File: lib/membrane/core/element/demand_controller.ex [hunk 2]
```diff
@@ -30,7 +30,7 @@ defmodule Membrane.Core.Element.DemandController do
     with {:ok, pad_data} when not pad_data.end_of_stream? <- PadModel.get_data(state, pad_ref),
          %State{playback: :playing} <- state do
       if pad_data.direction == :input do
-        raise("cannot snapshot atomic counter in input pad")
+        raise("cannot snapshot atomic counter in input pad #{inspect(pad_ref)}")
```
Comment (mat-hek) [inline lib/membrane/core/element/demand_controller.ex:33]:
```suggestion
        raise Membrane.ElementError, "Cannot snapshot atomic counter in input pad #{inspect(pad_ref)}"
```

**Review by mat-hek — APPROVED** [review]

**Comment by mat-hek** [comment]
@FelonEkonom BTW I recall you wanted to refactor `InputQueue` as well

**Comment by FelonEkonom** [comment]
@mat-hek I know, but then I decided that I would rather focus more on the code related to `DemandHandler`, because it was more messy and some changes, that I wanted to make in `InputQueue` (like deleting from it `AtomicDemand`) turned to be not a good ideas

**Pull request merged**